### PR TITLE
[SPARK-42533][CONNECT][Scala] Add ssl for Scala client

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Adding SSL encryption and access token support for Scala client

### Why are the changes needed?
To support basic client side encryption to protect data sent over the network.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests. Manual tests.

### CODE REVIEW

**Comment by zhenlineo** [comment]
cc @grundprinzip

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala [hunk 1]
```diff
@@ -117,6 +126,53 @@ object SparkConnectClient {
...
+    }
+
+    /**
+     * Disables the SSL. Throws exception if the token has been set.
+     *
+     * @return
+     *   this builder.
+     */
+    def disableSsl(): Builder = {
+      require(token.isEmpty, AUTH_TOKEN_ON_INSECURE_CONN_ERROR_MSG)
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:171]: So there is no way to disable this when the token is set?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala [hunk 2]
```diff
@@ -158,13 +214,14 @@ object SparkConnectClient {
...
-        } else {
-          // TODO(SPARK-41917): Support SSL and Auth tokens.
-          throw new UnsupportedOperationException(
-            "Parameters apart from user_id" +
-              " are currently unsupported.")
+        key match {
+          case URIParams.PARAM_USER_ID => userContextBuilder.setUserId(value)
+          case URIParams.PARAM_TOKEN => token(value)
+          case URIParams.PARAM_USE_SSL =>
+            if (true.toString.equalsIgnoreCase(value)) enableSsl() else disableSsl()
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:221]: `Boolean.valueOf()`?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala [hunk 3]
```diff
@@ -189,9 +245,54 @@ object SparkConnectClient {
     }
 
     def build(): SparkConnectClient = {
-      val channelBuilder = ManagedChannelBuilder.forAddress(host, port).usePlaintext()
-      val channel: ManagedChannel = channelBuilder.build()
+      def insecureCredentials(): ChannelCredentials = {
+        InsecureChannelCredentials.create()
+      }
+      val creds = isSslEnabled match {
+        case None => insecureCredentials()
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:252]: NIT, you can combine cases, e.g.: `case None | Some(false) => insecureCredentials()`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala (hunk 1 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:193]: @grundprinzip What's the correct action here? Should we allow users to turn off ssl when token is provided?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/MyTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+class MyTestSuite extends AnyFunSuite { // scalastyle:ignore funsuite
+
+  private var client: SparkConnectClient = _
+
+//  System.setProperty("javax.net.debug", "all")
+  test("Test ssl and token manually") {
+    client = SparkConnectClient
+      .builder()
+      .connectionString(
```
[the remark this answers is not in the document]
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/MyTestSuite.scala]: The client method name is not changed. But the `SparkSession.remote` method is added.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:193]: No, from the Python GRPC implementation it basically says that you cannot use a token on plaintext and we should follow this.

**Review by grundprinzip — APPROVED** [review]

**Comment by hvanhovell** [comment]
Merging to master/3.4

**Review by dongjoon-hyun — COMMENTED** [review]
Nice. Thank you all!

**Pull request closed without merging**